### PR TITLE
Minimal code completion support

### PR DESCRIPTION
**Pull request by lukechurch**
@devoncarew PTAL

This is the first very minimal cut of code completion integration.

### CODE REVIEW

File: lib/playground.dart [hunk 1]
```diff
@@ -534,10 +533,21 @@ class PlaygroundContext extends Context {
...
   Future<List<Completion>> complete(Editor editor) {
-    return new Future.value([
-      new Completion('one'),
-      new Completion('two'),
-      new Completion('three')
-    ]);
+
+    int offset = editor.document.indexFromPos(editor.document.cursor);
+
+    var sq =
```
Comment (devoncarew) [inline lib/playground.dart]: `request`?

File: lib/playground.dart [hunk 2]
```diff
@@ -534,10 +533,21 @@ class PlaygroundContext extends Context {
...
+        new SourceRequest()..source = editor.document.value
+        ..offset = offset;
+
+    return dartServices.complete(sq).then((cr) {
```
Comment (devoncarew) [inline lib/playground.dart]: `.then((response) {`?

**Comment by devoncarew** [comment]
lgtm! You may want to add a TODO to the DartCompleter, that it currently requests on each keystroke, rather than on each user completion request.

File: lib/playground.dart (hunk 1 above)
Comment (lukechurch) [inline lib/playground.dart]: done

File: lib/playground.dart (hunk 2 above)
Comment (lukechurch) [inline lib/playground.dart]: done

**Comment by lukechurch** [comment]
Thanks @devoncarew, wrt the comment I think you had already written one a couple of lines above :)

**Pull request merged**